hazelcast/src/main/java/com/hazelcast/internal/partition/impl/MigrationPlanner.java

```diff
@@ -0,0 +1,281 @@
+/*
+ * Copyright (c) 2008-2016, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.internal.partition.impl;
+
+import com.hazelcast.nio.Address;
+
+import java.util.Arrays;
+import java.util.HashSet;
+import java.util.Set;
+
+import static com.hazelcast.internal.partition.impl.InternalPartitionImpl.getReplicaIndex;
+
+/**
+ * Decides type and order of migrations that will move the state from current replica ownerships
+ * to the targeted replica ownerships. Migrations are planned in such a way that the current replica state will be moved to
+ * the targeted replica state one migration at a time. {@link MigrationDecisionCallback} implementation passed to the
+ * {@link MigrationPlanner#planMigrations(Address[], Address[], MigrationDecisionCallback)} is notified with the planned
+ * migrations. Planned migrations have a key property that they never decrease the available replica count of a partition.
+ */
+class MigrationPlanner {
+
+    private static final boolean ASSERTION_ENABLED = MigrationPlanner.class.desiredAssertionStatus();
+    private static final boolean TRACE = false;
+
+    interface MigrationDecisionCallback {
+        void migrate(Address source, int sourceCurrentReplicaIndex, int sourceNewReplicaIndex, Address destination,
+                     int destinationCurrentReplicaIndex, int destinationNewReplicaIndex);
+    }
+
+    private final Set<Address> verificationSet = new HashSet<Address>();
+
+    // checkstyle warnings are suppressed intentionally because the algorithm is followed easier within a single coherent method.
```

Comment:
I am aware of that, but having it in-lined in the code would really readability. 

This is usually achieved by having descriptive method names. However you decided for a single method handling all cases - I understand your choice, but one can feel kinda lost. 

A few words would do it. there is already a comment like this. see `// IT IS A SHIFT DOWN`